### PR TITLE
feat: provide git sha for cmd_all

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Display GIT SHA in CLI `risingwave -V`. This will always present. Original uses are unchanged (so `SELECT version()` can still be `unknown`). See comment for more info.

Also add some checks/fixes related with release.

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: ci/scripts/release.sh [hunk 1]
```diff
@@ -4,6 +4,7 @@
 set -euo pipefail
 
 REPO_ROOT=${PWD}
+export GIT_SHA=${BUILDKITE_COMMIT}
```
Comment (xxchan) [inline ci/scripts/release.sh:7]: Previously this is not set. So it's `unknown` for pre-built binary (it's only set for docker images).

```
dev=> select version();
                  version
--------------------------------------------
 PostgreSQL 8.3-RisingWave-0.19.0 (unknown)
```

File: ci/scripts/release.sh [hunk 2]
```diff
@@ -39,9 +40,15 @@ echo "--- Install aws cli"
...
+echo "--- Check risingwave release version"
 if [[ -n "${BUILDKITE_TAG+x}" ]]; then
-  toml set --toml-path Cargo.toml workspace.package.version ${BUILDKITE_TAG#*v}
+  CARGO_PKG_VERSION="$(toml get --toml-path Cargo.toml workspace.package.version)"
+  if [[ "${CARGO_PKG_VERSION}" != "${BUILDKITE_TAG#*v}" ]]; then
+    echo "CARGO_PKG_VERSION: ${CARGO_PKG_VERSION}"
+    echo "BUILDKITE_TAG: ${BUILDKITE_TAG}"
+    echo "CARGO_PKG_VERSION and BUILDKITE_TAG are not equal"
+    exit 1
+  fi
```
Comment (xxchan) [inline ci/scripts/release.sh:51]: Replaced update with check. This was only updated here, but not for docker image. So it was broken for docker image `v0.19.1` (It shows `v0.19.0` instead)

I think it's better to update it manually.

**Review by lmatz — APPROVED** [review]
LGTM, thanks!

File: ci/scripts/release.sh (hunk 2 above)
Comment (lmatz) [reply ci/scripts/release.sh:51]: Yep, let me update it

**Review by huangjw806 — APPROVED** [review]
LGTM!

**Review by arkbriar — APPROVED** [review]
LGTM!

File: ci/scripts/release.sh (hunk 2 above)
Comment (BugenZhao) [reply ci/scripts/release.sh:51]: +1

File: src/cmd_all/src/lib.rs [hunk 3]
```diff
@@ -16,3 +16,27 @@
...
+        // for each commit, so we only use it for the final binary.
+        let vergen_git_sha = env!("VERGEN_GIT_SHA");
+
+        // GIT_SHA is a normal environment variable. It's risingwave_common::GIT_SHA
+        // and is used in logs/version queries.
+        //
+        // Usually it's only provided by docker/binary releases (including nightly builds).
+        // We check it is the same as VERGEN_GIT_SHA when it's present.
+        if let Some(git_sha) = option_env!("GIT_SHA") {
+            assert!(git_sha.to_lowercase().contains(vergen_git_sha), "GIT_SHA ({git_sha}) mismatches VERGEN_GIT_SHA ({vergen_git_sha})");
```
Comment (BugenZhao) [inline src/cmd_all/src/lib.rs:33]: Can we make this a compile-time check?

**Review by BugenZhao — COMMENTED** [review]
Have we tested that whether this hurts incremental building for development?

**Comment by xxchan** [comment]
> So we have to build the final binary if we commit all staged files for development?

I tried `git commit --allow-empty` + `cargo run`. It's only 6s and I think it's acceptable.

File: src/cmd_all/src/lib.rs (hunk 3 above)
Comment (xxchan) [reply src/cmd_all/src/lib.rs]: image

File: src/cmd_all/src/bin/risingwave.rs [hunk 4]
```diff
@@ -106,10 +110,12 @@ impl Component {
 
 #[cfg_attr(coverage, no_coverage)]
 fn main() -> Result<()> {
+    let version: &'static str =
+        Box::leak(format!("{} ({})", CRATE_VERSION, VERGEN_GIT_SHA).into_boxed_str());
```
Comment (BugenZhao) [inline src/cmd_all/src/bin/risingwave.rs:114]: Can we also use `const_str` here?

**Review by BugenZhao — APPROVED** [review]
LGTM

**Pull request merged**